File: 12_simd_intrinsics/problems/02_poly_horner/poly_horner.f90
! 注: ベクトル型 (vector_size) は C/C++ 独自の拡張で Fortran には無い.
! Fortran では普通のループで Horner 法を書けば, コンパイラが自動的にSIMD化する.
! 多項式 p(x) = c(1) + c(2)*x + c(3)*x^2 + c(4)*x^3
program poly_horner
  implicit none
  character(len=32) :: arg
  integer(8) :: n, i, err
  integer :: k
  integer, parameter :: deg = 3
  real(8) :: c(0:deg) = (/ 1.0d0, 2.0d0, 3.0d0, 4.0d0 /)
  real(8) :: acc, r, d
  real(8), allocatable :: x(:), p(:)
  n = 64
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) n
  end if
  allocate(x(n), p(n))
  do i = 1, n
     x(i) = 0.001d0 * real(i - 1, 8)
  end do

  ! 各要素について Horner 法 acc = acc*x + c_k で多項式を評価する
  do i = 1, n
     ! TODO: Horner法 acc = acc*x + c_k で p(i) を求めよ.
  end do

  err = 0
  do i = 1, n
     r = c(deg)
     do k = deg - 1, 0, -1
        r = r * x(i) + c(k)
     end do
     d = p(i) - r
     if (d < -1.0d-9 .or. d > 1.0d-9) err = err + 1
  end do
  if (err == 0) then
     print "(a,f0.3,a,i0,a,f0.3)", "OK: p(1)=", p(1), ", p(", n, ")=", p(n)
  else
     print "(a,i0,a)", "NG: ", err, " 要素が不正"
  end if
end program poly_horner

File: 12_simd_intrinsics/problems/02_poly_horner/poly_horner.cpp
#include <cstdio>
#include <cstdlib>

enum { nl = 8 };   /* double 8つ = 512 bit のベクトル長 */
typedef double doublev __attribute__((vector_size(sizeof(double) * nl)));

/* 多項式 p(x) = c[0] + c[1]*x + c[2]*x^2 + c[3]*x^3 */
enum { deg = 3 };
static const double c[deg + 1] = { 1.0, 2.0, 3.0, 4.0 };

/* スカラ版 (検算用) */
static double poly_ref(double x) {
  double acc = c[deg];
  for (int k = deg - 1; k >= 0; k--) acc = acc * x + c[k];
  return acc;
}

int main(int argc, char ** argv) {
  /* n は簡単のため nl の倍数とする */
  long n = (argc > 1 ? atol(argv[1]) : 64);
  double * x = (double *)malloc(sizeof(double) * n);
  double * p = (double *)malloc(sizeof(double) * n);
  for (long i = 0; i < n; i++) x[i] = 0.001 * (double)i;

  /* nl 個ずつまとめて Horner 法 acc = acc*x + c_k で多項式を評価する */
  for (long i = 0; i < n; i += nl) {
    doublev xv = *(doublev *)&x[i];
    doublev acc;
    // TODO: Horner法 acc = acc*x + c_k をベクトルのまま計算して多項式を評価せよ.
    *(doublev *)&p[i] = acc;
  }

  long err = 0;
  for (long i = 0; i < n; i++) {
    double r = poly_ref(x[i]);
    double d = p[i] - r;
    if (d < -1e-9 || d > 1e-9) err++;
  }
  if (err == 0) printf("OK: p[0]=%.3f, p[%ld]=%.3f\n", p[0], n - 1, p[n - 1]);
  else          printf("NG: %ld 要素が不正\n", err);
  free(x); free(p);
  return 0;
}
